llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -0,0 +1,230 @@
+  bool runOnModule(Module &M) override {
+    LLVMContext &Context = M.getContext();
+
+    DataLayout DL(&M);
+    Type *Int8Ty = Type::getInt8Ty(Context);
+    Type *Int32Ty = Type::getInt32Ty(Context);
+    Type *PointerSizedIntTy = DL.getIntPtrType(Context);
+    Type *Int8PtrTy =  Type::getInt8PtrTy(Context);
+
+    // Create a global variable which will store the pointer to the heap memory
+    // allocated for the shadow stack.
+    Constant * GShadowStackPtr = M.getOrInsertGlobal("shadowstack_0", Int8PtrTy);
+    GlobalVariable *GVar = M.getNamedGlobal("shadowstack_0");
+    GVar->setInitializer(ConstantPointerNull::get(cast<PointerType>(Int8PtrTy)));
+
+    // We only need to create one shadow stack per module so we'll do this
+    // inside the module's entry point.
+    // YKFIXME: Investigate languages that don't have/use main as the first
+    // entry point.
+    Function *Main = M.getFunction("main");
+    if (Main == nullptr) {
+      Context.emitError("Unable to add shadow stack: could not find \"main\" function!");
+      return false;
+    }
+    Instruction *First = Main->getEntryBlock().getFirstNonPHI();
+    IRBuilder<> Builder(First);
+
+    // Now create some memory on the heap for the shadow stack.
+    FunctionCallee MF = M.getOrInsertFunction("malloc", Int8PtrTy, PointerSizedIntTy);
+    CallInst *Malloc = Builder.CreateCall(MF,
+        {ConstantInt::get(PointerSizedIntTy, SHADOW_STACK_SIZE)}, "");
+    Builder.CreateStore(Malloc, GShadowStackPtr);
+
+    Value *SSPtr;
+    for (Function &F : M) {
+      if (F.empty()) // skip declarations.
+        continue;
+
+      if (&F != Main) {
+        // At the top of each function in the module, load the heap pointer
+        // from the global shadow stack variable.
+        Builder.SetInsertPoint(F.getEntryBlock().getFirstNonPHI());
+        SSPtr = Builder.CreateLoad(Int8PtrTy, GShadowStackPtr);
+      } else {
+        SSPtr = cast<Value>(Malloc);
+      }
+
+      size_t Offset = 0;
+      // Remember which allocas were replaced, so we can remove them later in
+      // one swoop. Removing them here messes up the loop.
+      std::vector<Instruction *> RemoveAllocas;
+      for (BasicBlock &BB: F) {
+        for (Instruction &I: BB) {
+          if (isa<AllocaInst>(I)) {
+            // Replace allocas with pointers into the shadow stack.
+            AllocaInst &AI = cast<AllocaInst>(I);
+            if (isYkMTNewAlloca(&AI)) {
+              // The variable created by `yk_mt_new` will never be traced, so
+              // there's no need to store it on the shadow stack.
+              continue;
+            }
+            if (isa<StructType>(AI.getAllocatedType())) {
+              StructType *ST = cast<StructType>(AI.getAllocatedType());
+              // Some yk specific variables that will never be traced and thus
+              // can live happily on the normal stack.
+              if (ST->getName() == "YkCtrlPointVars" || ST->getName() == "struct.YkLocation") {
```

Comment:
This bit seems very fragile in the sense that it relies on other parts of the code transforming yk stuff? That fragility is probably inevitable, but we should consider at least documenting that fragility and, perhaps, adding an assert or two in case the other parts of the code change.

Reply:
All this does is keep yk specific variables on the normal stack. In theory they should also be fine on the shadow stack, though I was getting segfaults when that happened. Luckily for us, this is likely going away once we only move variables on the shadowstack that have their reference taken. I will add a comment.

Reply:
Try [033c98b](https://github.com/ykjit/ykllvm/pull/53/commits/033c98badf9bef2417cada97fb6c493c1f91ab75)